### PR TITLE
Consolidates the RdpdrBackend smartcard calls into a single function.

### PR DESCRIPTION
**Pull request by ibeckermayer**
Adds `EstablishContextCall` and `EstablishContextReturn` in the process.

### CODE REVIEW

File: crates/ironrdp-rdpdr/src/pdu/esc.rs [hunk 1]
```diff
@@ -464,6 +651,13 @@ pub mod rpce {
         fn size(&self) -> usize;
     }
 
+    pub trait HeaderlessDecode {
+        /// Decodes the instance from a buffer sans its [`StreamHeader`] and [`TypeHeader`].
+        fn decode(src: &mut ReadCursor<'_>) -> PduResult<Self>
+        where
+            Self: Sized;
+    }
```
[the remark this answers is not in the document]
Comment (ibeckermayer) [reply crates/ironrdp-rdpdr/src/pdu/esc.rs]: I went with dda4e024bbed2c66d85a97d84de6c64fdb8eb52c + 218ccaec83250414b6a2eb80b3f57715a6dae5ab to create symmetry with `HeaderlessEncode`

File: crates/ironrdp-rdpdr/src/pdu/esc.rs [hunk 2]
```diff
@@ -571,3 +798,84 @@ pub mod rpce {
         }
     }
 }
+
+pub mod ndr {
+    //! Request/response messages are nested structs with fields, encoded as NDR (network data
+    //! representation).
+    //!
+    //! Fixed-sized fields are encoded in-line as they appear in the struct.
```
[the remark this answers is not in the document]
Comment (ibeckermayer) [reply crates/ironrdp-rdpdr/src/pdu/esc.rs]: ChatGPT says "fixed-size" is correct so we'll run with that

File: crates/ironrdp-rdpdr/src/pdu/esc.rs [hunk 3]
```diff
@@ -497,12 +699,34 @@ pub mod rpce {
...
+                    "decode",
+                    "StreamHeader",
+                    "server returned big-endian data, parsing not implemented"
+                ))
+            }
+        }
+
         fn size() -> usize {
-            8
+            size_of::<u8>() + size_of::<u8>() + size_of::<u16>() + size_of::<u32>()
```
Comment (CBenoit) [inline crates/ironrdp-rdpdr/src/pdu/esc.rs:725]: The more I see this kind of code, the more I feel there is something better to do:
```suggestion
            1 /* Version */ + 1 /* Endianness */ + 2 /* CommonHeaderLength */ + 4 /* Filler */
```
It’s easier to double check with the Microsoft documentation:
image

And I don’t feel like `size_of::<u8>()` is really useful, it’s not like the size of `u8` was going to change and was hard to predict (that would be a problem for a network protocol in the first place). The name of the field is much more useful in my opinion.

I’m not asking that you change this everywhere immediately now, but I wanted to point that out and maybe change our approach for new code in the future.

cc @pacmancoder

**Review by CBenoit — APPROVED** [review]
LGTM, enabling auto-merge

**Pull request merged**